nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/PutHBaseRecord.java

```diff
@@ -0,0 +1,316 @@
+        getLogger().debug("Sent {} FlowFiles to HBase successfully in {} milliseconds", new Object[]{successes.size(), sendMillis});
+
+        for (FlowFile flowFile : successes) {
+            session.transfer(flowFile, REL_SUCCESS);
+            final String details = "Put " + columnsAdded.get(flowFile) + " cells to HBase";
+            session.getProvenanceReporter().send(flowFile, provenanceEvents.get(flowFile).get(0), details, sendMillis);
+            session.getProvenanceReporter().send(flowFile, provenanceEvents.get(flowFile).get(0), details, sendMillis);
+        }
+
+        session.commit();
+    }
+
+
+    @Override
+    protected PutFlowFile createPut(ProcessSession session, ProcessContext context, FlowFile flowFile) {
+        return null;
+    }
+
+    protected byte[] asBytes(String field, Object input) {
+        byte[] retVal = null;
+
+        if (input instanceof Number) {
+            if (input instanceof Float || input instanceof Double) {
+                retVal = clientService.toBytes(Double.parseDouble(input.toString()));
+            } else {
+                retVal = clientService.toBytes(Long.parseLong(input.toString()));
+            }
+        } else if (input instanceof Boolean) {
+            retVal = clientService.toBytes((Boolean)input);
+        } else if (input instanceof String) {
+            retVal = clientService.toBytes((String)input);
+        } else {
+            throw new RuntimeException(String.format("Could not identify type for field %", field));
+        }
+
+        return retVal;
+    }
+
+    protected byte[] asBytes(String field, RecordFieldType fieldType, Record record, boolean asString) {
+
+        byte[] retVal;
+
+        if (asString) {
+            retVal = Bytes.toBytes(record.getAsString(field));
+        } else {
+            switch (fieldType) {
+                case BOOLEAN:
+                    retVal = Bytes.toBytes(record.getAsBoolean(field));
+                    break;
+                case CHAR:
+                    retVal = Bytes.toBytes(record.getAsString(field));
+                    break;
+                case DOUBLE:
+                    retVal = Bytes.toBytes(record.getAsDouble(field));
+                    break;
+                case FLOAT:
+                    retVal = Bytes.toBytes(record.getAsFloat(field));
+                    break;
+                case INT:
+                    retVal = Bytes.toBytes(record.getAsInt(field));
+                    break;
+                case LONG:
+                    retVal = Bytes.toBytes(record.getAsLong(field));
+                    break;
+                default:
+                    retVal = Bytes.toBytes(record.getAsString(field));
+            }
+        }
+
+        return retVal;
+    }
+
+    protected List<PutFlowFile> createPuts(ProcessSession session, ProcessContext context, FlowFile flowFile, RecordReaderFactory recordParserFactory) {
```

Comment:
I think we should try to avoid reading the entire flow file into memory if possible... 

With PutHBaseJSON we had one JSON message per flow file, so the # of messages in memory at one time would be equal to the batch size. With the record oriented processors, the whole idea is to keep a lot of records together for efficiency, so we might have millions of records in a single flow file which could lead to memory issues if they were all read into memory.

A possible approach to deal with this could be the following... 

We could change the meaning of Batch Size in this processor, and instead of it being the number of flow files to grab, it could be the number of records to send to HBase in one call. So the processor grabs one flow file, creates a record reader and starts reading records up to batch size, sends the batch to HBase, and then starts a new batch. We can also keep track of the index of the last successful record, and if an error occurs at any point we then route the flow file to failure and add an attribute like "record.successful.index.<processor-id>". If the processor receives a flow file with this attribute already populated, then it will pick up from this index when sending records. 

Thoughts?